census2020/census2020.ipynb

```diff
@@ -0,0 +1,609 @@
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now, we will aggregate the data using the `ds.by()` parameter, which allows us to group the dataset by the `race` column. This will produce separate aggregates for each racial category which will now be colored using the specified `color_key`.\n",
+    "\n",
+    "Additionally, we will enable the **`dynspread=True`** parameter to dynamically adjust the size of rendered points. This ensures that areas with sparse data points remain visible, even when zooming out or when the data points are too densely packed. This dynamic spreading of points enhances visibility without overwhelming the plot, allowing us to capture both the subtle variations in rural areas and the dense clusters in urban regions. \n",
+    "\n",
+    "You can see how this works by zooming into specific regions of the map if you are working in a live Python environment. The plot will dynamically re-render, adjusting the point sizes and density to provide clearer visibility in the zoomed areas."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import holoviews as hv\n",
+    "import xyzservices.providers as xyz\n",
+    "from IPython.display import display\n",
+    "\n",
+    "label = hv.Tiles(xyz.CartoDB.PositronOnlyLabels())\n",
+    "\n",
+    "def plot_city(longitude_range, latitude_range, w=plot_width, h=plot_height, bgcolor='black'):\n",
+    "    \"\"\"Plots a map of the pop density aggregated by race within specified longitude and latitude ranges\"\"\"\n",
+    "    return df.hvplot.points('x', 'y', rasterize=True, dynspread=True, cnorm='eq_hist',\n",
+    "                            xlim=longitude_range, ylim=latitude_range,\n",
+    "                            cmap=color_key, bgcolor=bgcolor, colorbar=False, hover=False,\n",
+    "                            aggregator=ds.by('race'), width=w, height=h) * label"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%time\n",
+    "display(plot_city(*CITIES['USA']))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The result shows that the continental USA is overwhelmingly white in geographic terms, apart from some predominantly Hispanic regions along the Southern border, some regions with high densities of blacks in the Southeast, and a few isolated areas of Native Americans in the West.\n",
+    "\n",
+    "If you're working in a live Python environment, you can zoom into specific areas of the map, and the plot will automatically re-render to show more details at that zoom level. We can also zoom into specific regions using predefined coordinates:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%time\n",
+    "display(plot_city(*CITIES['New York City']))"
```

Comment:
My initial thinking then was that since the plots are now being called from a pre-defined function, it makes sense to time how long it takes to run but thinking about it further, there is no reason why the previous plots cannot also be timed. I guess I could also time them as well :)

Reply:
Something to think about mabe. Serving the app with `panel serve ...` executes the notebook content every time a visitor visits the page. Calling `display(plot(...))` forces the HoloViews plot to be generated, making the app longer to be served. In the streaming example of Demetris I mentioned `it not pn.state.served: ...` which could be used here inside a function to only execute `display()` when the app is not served.

Reply:
In these examples I've generally put timing information on any cell that takes a surprising amount of time. Sometimes the surprise is "how can Datashader be so fast", and other times it's answering "how long does that dataset take to load".

Reply:
Here's the streaming state bit: https://github.com/holoviz-topics/examples/pull/469/files#r1880352906